src/node_file.cc

```diff
@@ -2727,6 +2730,222 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+static bool FileURLToPathImpl(Environment* env,
+                       ada::url_aggregator* file_url,
+                       std::unique_ptr<std::string>& result_file_path) {
+  if (file_url->type != ada::scheme::FILE) {
+    env->isolate()->ThrowException(ERR_INVALID_URL_SCHEME(env->isolate()));
+
+    return false;
+  }
+
+  std::string_view pathname = file_url->get_pathname();
+#if _WIN_32
+  std::string pathname_escaped_slash;
+
+  for (size_t i = 0; (i + 2) < pathname.size(); ++i) {
+    pathname_escaped_slash += pathname[i];
+
+    if (pathname[i] == '\\') pathname_escaped_slash += '\\';
+
+    if (pathname[i] != '%') continue;
+
+    char third = pathname[i + 2] | 0x20;
+
+    bool is_slash = pathname[i + 1] == '2' && third == 102;         // 2f 2F /
+    bool is_forward_slash = pathname[i + 1] == '5' && third == 99;  // 5c 5C \
+
+    if (!is_slash && !is_forward_slash) continue;
+
+    env->isolate()->ThrowException(ERR_INVALID_FILE_URL_PATH(
+        env->isolate(),
+        "File URL path must not include encoded \\ or / characters"));
+
+    return false;
+  }
+
+  std::string decoded_pathname = ada::unicode::percent_decode(
+      std::string_view(pathname_escaped_slash), pathname_escaped_slash.size());
+
+  if (hostname.size() > 0) {
+    // If hostname is set, then we have a UNC path
+    // Pass the hostname through domainToUnicode just in case
+    // it is an IDN using punycode encoding. We do not need to worry
+    // about percent encoding because the URL parser will have
+    // already taken care of that for us. Note that this only
+    // causes IDNs with an appropriate `xn--` prefix to be decoded.
+    *result_file_path =
+        "\\\\" + ada::unicode::to_unicode(hostname) + decoded_pathname;
+
+    return true;
+  }
+
+  char letter = decoded_pathname[1] | 0x20;
+  char sep = decoded_pathname[2];
+
+  // a..z A..Z
+  if (letter < 'a' || letter > 'z' || sep != ':') {
+    env->isolate()->ThrowException(ERR_INVALID_FILE_URL_PATH(
+        env->isolate(), "File URL path must be absolute"));
+
+    return false;
+  }
+
+  *result_file_path = decoded_pathname.substr(1);
+
+  return true;
+#else
+  std::string_view hostname = file_url->get_hostname();
+
+  if (hostname.size() > 0) {
+    std::string error_message =
+        std::string("File URL host must be \"localhost\" or empty on ") +
+        std::string(per_process::metadata.platform);
+    env->isolate()->ThrowException(
+        ERR_INVALID_FILE_URL_HOST(env->isolate(), error_message.c_str()));
+
+    return false;
+  }
+
+  int first_percent = 0;
+
+  for (std::string::size_type i = 0; (i + 2) < pathname.size(); ++i) {
+    if (pathname[i] != '%') continue;
+
+    if (first_percent == 0) {
+      first_percent = i;
+    }
+
+    if (pathname[i + 1] != '2' || (pathname[i + 2] | 0x20) != 102) continue;
```

Comment:
For readability, i think it is better to just have a single if statement for false case.